codalab/lib/bundle_cli.py

```diff
@@ -1823,6 +1824,53 @@ def do_rm_command(self, args):
             for uuid in deleted_uuids:
                 print(uuid, file=self.stdout)
 
+    @Commands.command(
+        'ancestors',
+        help='Print ancestors.',
+        arguments=(
+            Commands.Argument(
+                'bundle_spec', help=BUNDLE_SPEC_FORMAT, nargs='+', completer=BundlesCompleter
+            ),
+            Commands.Argument(
+                '-w',
+                '--worksheet-spec',
+                help='Operate on this worksheet (%s).' % WORKSHEET_SPEC_FORMAT,
+                completer=WorksheetsCompleter,
+            ),
+        ),
+    )
+
+    def do_ancestors_command(self, args):
+        def print_ancestor(format_str, indent_level, bundle_name, bundle_id):
+            print(format_str.format(
+                    " "*indent_level,
+                    bundle_name,
+                    worksheet_util.apply_func(UUID_POST_FUNC, bundle_id)), file=self.stdout)
+
+        bundle_stack = []
+        processed_bundles = set()
+        args.bundle_spec = spec_util.expand_specs(args.bundle_spec)
+        client, worksheet_uuid = self.parse_client_worksheet_uuid(args.worksheet_spec)
+
+        # Get the initial bundle uuids and enqueue them for processing
+        bundle_uuids = self.target_specs_to_bundle_uuids(client, worksheet_uuid, args.bundle_spec)
+        bundle_stack.extend([(uuid, 0) for uuid in bundle_uuids])
+
+        # Traversal through all ancestors, avoid printing entire tree on repeated worksheets
+        while (len(bundle_stack) != 0):
+            curr_bundle_uuid, level = bundle_stack.pop()
+            # Assume only 1 bundle in bundle array
+            curr_bundle = client.fetch('bundles', params={'specs': curr_bundle_uuid})[0]
+            if curr_bundle_uuid in processed_bundles:
+                print_ancestor('{}- {}({})...', level, curr_bundle['metadata']['name'], curr_bundle_uuid)
```

Comment:
I'd do this: `print(render(curr_bundle) + ('...' if curr_bundle_uuid in processed_bundles else ''))`

Reply:
Is the render method you suggest equivalent the print_ancestor function (should it take the full list of params)?

Reply:
The main thing I'm trying to get at is trying to have more code sharing (which means minimizing the arguments - I guess you need `level`).